lib/actions/api.js

```diff
@@ -574,14 +694,63 @@ function createQueryAction (endpoint, responseAction, errorAction, options) {
       return dispatch(errorAction(err))
     }
 
-    if (options && typeof options.rewritePayload === 'function') {
+    if (typeof options.rewritePayload === 'function') {
       dispatch(responseAction(options.rewritePayload(payload)))
     } else {
       dispatch(responseAction(payload))
     }
 
-    if (options && typeof options.postprocess === 'function') {
+    if (typeof options.postprocess === 'function') {
       options.postprocess(payload, dispatch, getState)
     }
   }
 }
+
+function createGraphQLQueryAction (query, variables, responseAction, errorAction, options) {
+  const endpoint = `index/graphql`
+  const fetchOptions = {
+    method: 'POST',
+    body: JSON.stringify({ query, variables }),
+    headers: { 'Content-Type': 'application/json' }
+  }
+  return createQueryAction(
+    endpoint,
+    responseAction,
+    errorAction,
+    { ...options, fetchOptions }
+  )
+}
+
+/**
+ * Update the browser/URL history with new parameters
+ * NOTE: This has not been tested for profile-based journeys.
+ */
+export function setUrlSearch (params, replaceCurrent = false) {
+  return function (dispatch, getState) {
+    const base = getState().router.location.pathname
+    const path = `${base}?${qs.stringify(params)}`
+    if (replaceCurrent) dispatch(replace(path))
+    else dispatch(push(path))
+  }
+}
+
+/**
+ * Update the OTP Query parameters in the URL and ensure that the active search
+ * is set correctly. Leaves any other existing URL parameters (e.g., UI) unchanged.
+ */
+export function updateOtpUrlParams (otpState, searchId) {
+  const otpParams = getRoutingParams(otpState, true)
+  return function (dispatch, getState) {
+    const params = {}
+    // Get all OTP-specific params, which will be retained unchanged in the URL
+    const urlParams = getUrlParams()
+    Object.keys(urlParams)
+      .filter(key => key.indexOf('_') !== -1)
+      .forEach(key => { params[key] = urlParams[key] })
+    params.ui_activeSearch = searchId
+    // Assumes this is a new search and the active itinerary should be reset.
+    params.ui_activeItinerary = 0
+    // Merge in the provided OTP params and update the URL
+    dispatch(setUrlSearch(Object.assign(params, otpParams)))
+  }
+}
```

Comment:
This file is getting long. Let's break it up in the next round of feature development.